File: n8n-api/liderix_api/schemas/crm.py
from __future__ import annotations

from datetime import datetime
from typing import Optional, List, Generic, TypeVar
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator

from liderix_api.enums import (
    CRMContactStatus,
    CRMContactPriority,
    CRMContactSource,
    CRMDealStage,
)


class CRMContactBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    email: EmailStr
    phone: Optional[str] = Field(default=None, max_length=50)
    company: Optional[str] = Field(default=None, max_length=200)
    position: Optional[str] = Field(default=None, max_length=200)
    status: CRMContactStatus = CRMContactStatus.LEAD
    priority: CRMContactPriority = CRMContactPriority.MEDIUM
    source: CRMContactSource = CRMContactSource.WEBSITE
    value: Optional[float] = None
    last_contact: Optional[datetime] = None
    next_follow_up: Optional[datetime] = None
    notes: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    org_id: Optional[UUID] = None

    @field_validator("name")
    @classmethod
    def _strip_name(cls, v: str) -> str:
        return v.strip()

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class CRMContactCreate(CRMContactBase):
    pass


class CRMContactUpdate(BaseModel):
    name: Optional[str] = Field(default=None, max_length=200)
    email: Optional[EmailStr] = None
    phone: Optional[str] = Field(default=None, max_length=50)
    company: Optional[str] = Field(default=None, max_length=200)
    position: Optional[str] = Field(default=None, max_length=200)
    status: Optional[CRMContactStatus] = None
    priority: Optional[CRMContactPriority] = None
    source: Optional[CRMContactSource] = None
    value: Optional[float] = None
    last_contact: Optional[datetime] = None
    next_follow_up: Optional[datetime] = None
    notes: Optional[str] = None
    tags: Optional[List[str]] = None

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class CRMContactRead(BaseModel):
    id: UUID
    org_id: UUID
    name: str
    email: EmailStr
    phone: Optional[str] = None
    company: Optional[str] = None
    position: Optional[str] = None
    status: CRMContactStatus
    priority: CRMContactPriority
    source: CRMContactSource
    value: Optional[float] = None
    last_contact: Optional[datetime] = None
    next_follow_up: Optional[datetime] = None
    notes: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, extra="forbid")


T = TypeVar("T")


class CRMContactListResponse(BaseModel, Generic[T]):
    items: List[CRMContactRead]
    page: int
    page_size: int
    total: int

    model_config = ConfigDict(extra="forbid")


class CRMDealBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    contact_id: Optional[UUID] = None
    company: Optional[str] = Field(default=None, max_length=200)
    amount: float = Field(default=0.0, ge=0)
    stage: CRMDealStage = CRMDealStage.PROSPECTING
    probability: int = Field(default=0, ge=0, le=100)
    expected_close_date: Optional[datetime] = None
    org_id: Optional[UUID] = None

    @field_validator("title")
    @classmethod
    def _strip_title(cls, v: str) -> str:
        return v.strip()

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class CRMDealCreate(CRMDealBase):
    pass


class CRMDealUpdate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=200)
    contact_id: Optional[UUID] = None
    company: Optional[str] = Field(default=None, max_length=200)
    amount: Optional[float] = Field(default=None, ge=0)
    stage: Optional[CRMDealStage] = None
    probability: Optional[int] = Field(default=None, ge=0, le=100)
    expected_close_date: Optional[datetime] = None

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class CRMDealRead(BaseModel):
    id: UUID
    org_id: UUID
    title: str
    contact_id: Optional[UUID] = None
    contact_name: Optional[str] = None
    company: Optional[str] = None
    amount: float
    stage: CRMDealStage
    probability: int
    expected_close_date: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, extra="forbid")


class CRMDealListResponse(BaseModel, Generic[T]):
    items: List[CRMDealRead]
    page: int
    page_size: int
    total: int

    model_config = ConfigDict(extra="forbid")

File: apps/web-enterprise/src/lib/api/crm.ts
import { api } from "@/lib/api/config"
import { CompanyAPI } from "@/lib/api/company"

export type CRMContactStatus = "lead" | "prospect" | "customer" | "inactive"
export type CRMContactPriority = "low" | "medium" | "high"
export type CRMContactSource = "website" | "referral" | "cold_outreach" | "event" | "social"
export type CRMDealStage =
  | "prospecting"
  | "qualification"
  | "proposal"
  | "negotiation"
  | "closed_won"
  | "closed_lost"

export interface CRMContact {
  id: string
  org_id: string
  name: string
  email: string
  phone?: string | null
  company?: string | null
  position?: string | null
  status: CRMContactStatus
  priority: CRMContactPriority
  source: CRMContactSource
  value?: number | null
  last_contact?: string | null
  next_follow_up?: string | null
  notes?: string | null
  tags: string[]
  created_at: string
  updated_at: string
}

export interface CRMDeal {
  id: string
  org_id: string
  title: string
  contact_id?: string | null
  contact_name?: string | null
  company?: string | null
  amount: number
  stage: CRMDealStage
  probability: number
  expected_close_date?: string | null
  created_at: string
  updated_at: string
}

interface CRMListResponse<T> {
  items: T[]
  page: number
  page_size: number
  total: number
}

async function getDefaultOrgId(): Promise<string | null> {
  const org = await CompanyAPI.getCurrentCompany()
  return org?.id ?? null
}

export async function getContacts(orgId?: string): Promise<CRMContact[]> {
  const resolvedOrgId = orgId ?? (await getDefaultOrgId())
  const query = resolvedOrgId ? `?org_id=${encodeURIComponent(resolvedOrgId)}` : ""
  const { data } = await api.get<CRMListResponse<CRMContact>>(`/crm/contacts${query}`)
  return data.items ?? []
}

export async function createContact(payload: Omit<CRMContact, "id" | "created_at" | "updated_at" | "org_id"> & { org_id?: string }): Promise<CRMContact> {
  const orgId = payload.org_id ?? (await getDefaultOrgId())
  const { data } = await api.post<CRMContact>("/crm/contacts", {
    ...payload,
    org_id: orgId ?? undefined,
  })
  return data
}

export async function updateContact(contactId: string, payload: Partial<CRMContact>): Promise<CRMContact> {
  const { data } = await api.patch<CRMContact>(`/crm/contacts/${contactId}`, payload)
  return data
}

export async function deleteContact(contactId: string): Promise<void> {
  await api.delete(`/crm/contacts/${contactId}`)
}

export async function getDeals(orgId?: string): Promise<CRMDeal[]> {
  const resolvedOrgId = orgId ?? (await getDefaultOrgId())
  const query = resolvedOrgId ? `?org_id=${encodeURIComponent(resolvedOrgId)}` : ""
  const { data } = await api.get<CRMListResponse<CRMDeal>>(`/crm/deals${query}`)
  return data.items ?? []
}

export async function createDeal(payload: Omit<CRMDeal, "id" | "created_at" | "updated_at" | "org_id" | "contact_name"> & { org_id?: string }): Promise<CRMDeal> {
  const orgId = payload.org_id ?? (await getDefaultOrgId())
  const { data } = await api.post<CRMDeal>("/crm/deals", {
    ...payload,
    org_id: orgId ?? undefined,
  })
  return data
}

export async function updateDeal(dealId: string, payload: Partial<CRMDeal>): Promise<CRMDeal> {
  const { data } = await api.patch<CRMDeal>(`/crm/deals/${dealId}`, payload)
  return data
}

export async function deleteDeal(dealId: string): Promise<void> {
  await api.delete(`/crm/deals/${dealId}`)
}
